File: src/main/java/com/derongan/minecraft/mineinabyss/GUI/GUICommandExecutor.java
package com.derongan.minecraft.mineinabyss.GUI;

import com.derongan.minecraft.mineinabyss.AbyssContext;
import com.derongan.minecraft.mineinabyss.MineInAbyss;
import org.bukkit.Material;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GUICommandExecutor implements CommandExecutor {
    private AbyssContext context;

    public GUICommandExecutor(AbyssContext context) {
        this.context = context;
    }

    @Override
    public boolean onCommand(CommandSender commandSender, Command command, String label, String[] args) {
        if (commandSender instanceof Player) {
            Player player = (Player) commandSender;

            if (label.equals("stats")) {
                new StatsGUI(player, JavaPlugin.getPlugin(MineInAbyss.class)).show(player);
                return true;
            } else if (label.equals("start") && player.hasPermission("mineinabyss.start")) {
                new GondolaGUI(player, JavaPlugin.getPlugin(MineInAbyss.class)).show(player);
                return true;
            } else if (label.equals("creategondolaspawn") && player.hasPermission("mineinabyss.createspawn")) {
                FileConfiguration spawnLocConfig = context.getConfigManager().getSpawnLocConfig();
                List<Map<?, ?>> spawns = spawnLocConfig.getMapList(GondolaGUI.SPAWN_KEY);
                ItemStack displayItem = player.getInventory().getItemInMainHand().clone();

                if (displayItem.getType().equals(Material.AIR))
                    displayItem = new ItemStack(Material.GRASS_BLOCK);

                ItemMeta meta = displayItem.getItemMeta();
                if (args.length > 0) meta.setDisplayName(args[0]);
                meta.addItemFlags(ItemFlag.HIDE_ATTRIBUTES);
                meta.addItemFlags(ItemFlag.HIDE_UNBREAKABLE);
                displayItem.setItemMeta(meta);

                Map<String, Object> map = new HashMap<>();
                map.put("location", player.getLocation());
                map.put("display-item", displayItem);
                if (args.length >= 2)
                    map.put("cost", args[1]);

                spawns.add(map);
                spawnLocConfig.set(GondolaGUI.SPAWN_KEY, spawns);
                context.getConfigManager().saveSpawnLocConfig();
                player.sendMessage("Created spawn");
                return true;
            }
        }

        return false;
    }
}

File: src/main/java/com/derongan/minecraft/mineinabyss/commands/GUICommandExecutor.kt
package com.derongan.minecraft.mineinabyss.commands

import com.derongan.minecraft.deeperworld.services.WorldManager
import com.derongan.minecraft.mineinabyss.configuration.MIAConfig
import com.derongan.minecraft.mineinabyss.configuration.SpawnLocation
import com.derongan.minecraft.mineinabyss.configuration.SpawnLocationsConfig
import com.derongan.minecraft.mineinabyss.ecs.components.DescentContext
import com.derongan.minecraft.mineinabyss.ecs.components.pins.ActivePins
import com.derongan.minecraft.mineinabyss.ecs.systems.OrthReturnSystem
import com.derongan.minecraft.mineinabyss.gui.pins.PinMenu
import com.derongan.minecraft.mineinabyss.mineInAbyss
import com.derongan.minecraft.mineinabyss.player.openHubStorage
import com.mineinabyss.geary.ecs.prefab.PrefabKey
import com.mineinabyss.geary.minecraft.access.toGeary
import com.mineinabyss.guiy.inventory.guiy
import com.mineinabyss.idofront.commands.arguments.intArg
import com.mineinabyss.idofront.commands.arguments.stringArg
import com.mineinabyss.idofront.commands.execution.ExperimentalCommandDSL
import com.mineinabyss.idofront.commands.execution.IdofrontCommandExecutor
import com.mineinabyss.idofront.commands.execution.stopCommand
import com.mineinabyss.idofront.commands.extensions.actions.playerAction
import com.mineinabyss.idofront.items.editItemMeta
import com.mineinabyss.idofront.messaging.color
import com.mineinabyss.idofront.messaging.error
import com.mineinabyss.idofront.messaging.info
import com.mineinabyss.idofront.messaging.success
import com.mineinabyss.idofront.serialization.toSerializable
import org.bukkit.Material

@ExperimentalCommandDSL
object GUICommandExecutor : IdofrontCommandExecutor() {
    override val commands = commands(mineInAbyss) {
        ("mineinabyss" / "mia") command@{
            "pin" {
                "add" {
                    val key by arg<PrefabKey> {
                        parseBy { PrefabKey.of(passed) }
                    }

                    playerAction {
                        val pins = player.toGeary().getOrSetPersisting { ActivePins() }
                        pins.add(key)
                    }
                }
            }

            "pins" {
                playerAction {
                    guiy { PinMenu(player) }
                }
            }

            "start" {
                playerAction {
                    player.toGeary().apply {
                        if (has<DescentContext>())
                            this@command.stopCommand("You are already ingame!\nYou can leave using /stopdescent")
                        setPersisting(DescentContext())
                    }
//                    GondolaGUI(player).show(player)
                }
            }
            "stats" {
                "subcommand" {

                }
                playerAction {
                    // StatsGUI(player).show(player)
                }
            }
            "storage" {
                playerAction {
                    if (MIAConfig.data.hubSection == WorldManager.getSectionFor(player.location))
                        player.openHubStorage()
                    else
                        sender.error("You are not in the hub area")
                }
            }
            "stopdescent" {
                playerAction {
                    with(player.toGeary()) {
                        val descent = get<DescentContext>()
                            ?: this@command.stopCommand("You are not currently ingame!\nStart by using /start")
                        if (!descent.confirmedLeave) {
                            descent.confirmedLeave = true
                            sender.info(
                                """
                        &cYou are about to leave the game!!!
                        &lYour progress will be lost&r&c, but any xp and money you earned will stay with you.
                        Type /stopdescent again to leave
                        """.trimIndent().color()
                            )
                        } else {
                            player.health = 0.0
                            OrthReturnSystem.removeDescentContext(player)
                        }
                    }
                }
            }
            "creategondolaspawn" {
                val displayName by stringArg { default = "" }
                val cost by intArg { default = 0 }

                playerAction {
                    SpawnLocationsConfig.data.spawns
                    val displayItem = player.inventory.itemInMainHand

                    if (displayItem.type == Material.AIR)
                        displayItem.type = Material.GRASS_BLOCK

                    displayItem.editItemMeta {
                        if (displayName.isNotEmpty())
                            setDisplayName(displayName)
//                    addItemFlags(ItemFlag.HIDE_ATTRIBUTES) //TODO probably better to add these tags to the serialized items
//                    addItemFlags(ItemFlag.HIDE_UNBREAKABLE)
                    }

                    SpawnLocationsConfig.data.spawns.add(
                        SpawnLocation(
                            location = player.location,
                            displayItem = displayItem.toSerializable(),
                            cost = cost
                        )
                    )
                    SpawnLocationsConfig.queueSave()

                    sender.success("Created spawn")
                }
            }
        }
    }
}
